### PR TITLE
feat: implement body mixin on Readable

### PR DESCRIPTION
**Pull request by ronag**
Port over PR from node core. It will not be possible to land
these changes on core due to ecosystem breakage.

Refs: https://github.com/nodejs/node/pull/39520

### CODE REVIEW

File: lib/api/readable.js [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        this[kBody].getReader() // Ensure stream is locked.
+        assert(this[kBody].locked)
+      }
+    }
+    return this[kBody]
+  }
+}
+
+// https://streams.spec.whatwg.org/#readablestream-locked
+function isLocked (self) {
```
Comment (Ethan-Arrowood) [inline lib/api/readable.js:142]: Any particular reason these methods are not apart of the class itself? seems like everything here would be fine using `this`

File: lib/api/readable.js [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+const { RequestAbortedError, NotSupportedError } = require('../core/errors')
+
+let Blob
+
+const kConsume = Symbol('kConsume')
+const kReading = Symbol('kReading')
+const kBody = Symbol('kBody')
+const kAbort = Symbol('abort')
+
+module.exports = class BodyReadable extends Readable {
```
Comment (Ethan-Arrowood) [inline lib/api/readable.js:16]: I didn't get a chance to read through the old PR, but why not use the experimental WebStreams stuff? https://nodejs.org/api/webstreams.html#webstreams_class_readablestream

My personal plan for undici-fetch was to build on top of that

Comment (ronag) [reply lib/api/readable.js:16]: > I didn't get a chance to read through the old PR, but why not use the experimental WebStreams stuff?

Performance and ergonomics. Also making sure it doesn't break the existing API.

File: lib/api/readable.js (hunk 1 above)
Comment (ronag) [reply lib/api/readable.js:142]: It's just a simple way to make things private without having to use symbols.

File: lib/api/readable.js (hunk 2 above)
Comment (ronag) [reply lib/api/readable.js:16]: Also I guess it makes it easier for upstream libraries to also implement fetch body mixin.

File: lib/api/readable.js (hunk 1 above)
Comment (Ethan-Arrowood) [reply lib/api/readable.js:142]: Ahh good point. I usually only care to enforce this with TypeScript types. I see no reason to change this here though. I really like this implementation; great work!

File: lib/api/readable.js (hunk 2 above)
Comment (Ethan-Arrowood) [reply lib/api/readable.js:16]: Very smart! LGTM

**Review by Ethan-Arrowood — APPROVED** [review]
LGTM - do you want to make the type updates as well? Shouldn't be hard for this change

File: lib/api/readable.js [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    // Loop
+  }
+}
+
+function consumeEnd (consume) {
+  const { type, body, resolve, stream, length } = consume
+
+  try {
+    if (type === 'text') {
+      resolve(body.join(''))
```
Comment (szmarczak) [inline lib/api/readable.js:226]: How the performance differs with `Buffer.concat(body).toString()`?

Comment (ronag) [reply lib/api/readable.js:226]: No idea.

Comment (ronag) [reply lib/api/readable.js:226]: Probably creates an unnecessary memory buffer that way.

Comment (szmarczak) [reply lib/api/readable.js:226]: Looks like `.join('')` is much faster:

```js
const chunks = [];

for (let i = 0; i < 5_000; i++) {
	chunks.push(Buffer.allocUnsafe(1024 * 64));
}

console.time('join');
chunks.join('');
console.timeEnd('join');

console.time('concat');
Buffer.concat(chunks).toString();
console.timeEnd('concat');
```

```
join: 463.98ms
concat: 771.028m
```

**Review by mcollina — COMMENTED** [review]
Code looks good. I would recommend we mention what methods will be available in the docs. The .json() method might also deserve a special mention in the README

**Comment by ronag** [comment]
Fixed

File: lib/api/readable.js [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+    if (err) {
+      this[kAbort]()
+    }
+
+    return super.destroy(err)
+  }
+
+  emit (ev, ...args) {
```
Comment (mcollina) [inline lib/api/readable.js:50]: Can you check if this PR impact the performance of undici?

**Comment by ronag** [comment]
master
```
[bench:run] │ Tests               │ Samples │        Result │ Tolerance │ Difference with slowest │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ http - no keepalive │      10 │  6.39 req/sec │  ± 0.76 % │                       - │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ http - keepalive    │      10 │  7.14 req/sec │  ± 1.22 % │               + 11.87 % │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ undici - pipeline   │      10 │ 67.28 req/sec │  ± 1.88 % │              + 953.57 % │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ undici - request    │      10 │ 68.88 req/sec │  ± 1.87 % │              + 978.72 % │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ undici - stream     │      10 │ 70.55 req/sec │  ± 1.32 % │             + 1004.83 % │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ undici - dispatch   │      10 │ 75.16 req/sec │  ± 2.16 % │             + 1076.97 % │
```

PR
```js
[bench:run] │ Tests               │ Samples │        Result │ Tolerance │ Difference with slowest │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ http - no keepalive │      10 │  6.43 req/sec │  ± 0.98 % │                       - │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ http - keepalive    │      10 │  6.85 req/sec │  ± 1.93 % │                + 6.42 % │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ undici - pipeline   │      10 │ 63.27 req/sec │  ± 2.41 % │              + 883.47 % │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ undici - stream     │      10 │ 67.86 req/sec │  ± 1.31 % │              + 954.73 % │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ undici - request    │      10 │ 68.78 req/sec │  ± 2.16 % │              + 969.07 % │
[bench:run] |─────────────────────|─────────|───────────────|───────────|─────────────────────────|
[bench:run] │ undici - dispatch   │      10 │ 73.52 req/sec │  ± 1.72 % │             + 1042.76 % │
```

**Comment by ronag** [comment]
@mcollina doesn't seem to have any perf regression.

File: lib/api/readable.js (hunk 4 above)
Comment (ronag) [reply lib/api/readable.js:50]: Doesn't seem like that. See PR comment.

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**